lib/web_ui/dev/roll_fallback_fonts.dart

```diff
@@ -509,3 +494,354 @@ bool _checkForLicenseAttribution(Uint8List fontBytes) {
+  _FontSet(this.fonts);
+
+  /// The number of [_Font]s in this set.
+  int get length => fonts.length;
+
+  /// The members of this set.
+  final List<_Font> fonts;
+
+  /// Number of unicode ranges that are supported by this set of fonts.
+  int rangeCount = 0;
+
+  /// The serialization order of this set. This index is assigned after building
+  /// all the sets.
+  late final int index;
+
+  static int orderByDecreasingRangeCount(_FontSet a, _FontSet b) {
+    final int r = b.rangeCount.compareTo(a.rangeCount);
+    if (r != 0) {
+      return r;
+    }
+    return orderByLexicographicFontIndexes(a, b);
+  }
+
+  static int orderByLexicographicFontIndexes(_FontSet a, _FontSet b) {
+    for (int i = 0; i < a.length && i < b.length; i++) {
+      final int r = _Font.compare(a.fonts[i], b.fonts[i]);
+      if (r != 0) {
+        return r;
+      }
+    }
+    assert(a.length != b.length); // _FontSets are canonical.
+    return a.length - b.length;
+  }
+
+  @override
+  String toString() {
+    return description();
+  }
+
+  String description() {
+    return fonts.map((_Font font) => font.shortName).join(', ');
+  }
+}
+
+/// A trie node [1] used to find the canonical _FontSet.
+///
+/// [1]: https://en.wikipedia.org/wiki/Trie
+class _TrieNode {
+  final Map<_Font, _TrieNode> _children = <_Font, _TrieNode>{};
+  _FontSet? fontSet;
+
+  _TrieNode insertAtRoot(Iterable<_Font> fonts) {
+    _TrieNode node = this;
+    for (final _Font font in fonts) {
+      node = node._children[font] ??= _TrieNode();
+    }
+    return node;
+  }
+}
+
+/// Computes the Dart source code for the encoded data structures used by the
+/// fallback font selection algorithm.
+///
+/// The data structures allow the fallback font selection algorithm to quickly
+/// determine which fonts support a given code point. The structures are
+/// essentially a map from a code point to a set of fonts that support that code
+/// point.
+///
+/// The universe of code points is partitioned into a set of subsets, or
+/// components, where each component contains all the code points that are in
+/// exactly the same set of fonts. A font can be considered to be a union of
+/// some subset of the components and may share components with other fonts.  A
+/// `_FontSet` is used to represent a component and the set of fonts that use
+/// the component. One way to visualize this is as a Venn diagram. The fonts are
+/// the overlapping circles and the components are the spaces between the lines.
+///
+/// The emitted data structures are
+///
+///  (1) A list of sets of fonts.
+///  (2) A list of code point ranges mapping to an index of list (1).
+///
+/// Each set of fonts is represented as a list of font indexes. The indexes are
+/// always increasing so the delta is stored. The stored value is biased by -1
+/// (i.e.  `delta - 1`) since a delta is never less than 1. The deltas are STMR
+/// encoded.
+///
+/// The list of code point ranges is complete, covering every code point. There
```

Comment:
docs nit: this point wasn't clear to me until I started reading the algorithm and saw that the empty FontSet covered the uncovered ranges. Maybe change this paragraph to highlight the importance of the empty set of fonts in the first sentence?

Reply:
Done.